### PR TITLE
"Basic Points For Badges" Components

### PR DESCRIPTION
**Pull request by chrisnowak**
## Overview

This introduces a fairly basic set of components for a points and badges system.  The overall design here is already IRB-approved; while this definitely has some limitations (notably mostly client-side calculation), I think it will work nicely as a drop-in component for studies who just want to add a little "gamification" to their project.

Screenshot 2024-05-28 at 3 34 22 PM
Screenshot 2024-05-28 at 3 26 22 PM

There's a handful of components here:
- BasicPointsForBadges is the core component.  This basically:
  - Gets the current state of all "activities" and their points
  - Calculates what the updated points should be based on newly completed activities
  - Animates the progress bar to the new total
  - If we reached the total number of points per badge, displays a modal congratulating the participant
- DailyDataGoal
  - Can be used entirely independently of this module; basically just lets you specify a type of daily data and a goal threshold that you want the participant to reach.  You can also specify messages that show at certain thresholds of that goal.
- BasicBadges
  - Shows the badges the participant has collected.  
- NewBadgeView
  - What is displayed when a new badge is earned; displays the latest badge and the point total.  (If multiple badges are earned at once, it'll just display the latest)

There's also a supporting cast of presentational components:
- BasicBadge
  - Presentational component for the badge that is used in the BasicBadges component and in NewBadgeView
- Grid / Grid.Column
  - I decided to tackle a very basic grid component based on what i've seen in other UI frameworks to be able to have a left/right layout.  
  - In particular this was because the design requires DailyDataGoal components to be left/right aligned.  I could have created a "DailyDataGoals" component but it felt preferable to have someone be able to drag/drop these individually via view builder
- Divider
  - Instead of optionally attaching border properties to all these components it feels easy to just have a "divider" component

### Points and Badges Calculation
The calculations are all performed client side at the moment (with the option of also getting points from a custom field); which of course has a bunch of downsides (can't send notifications of goals being completed passively, not advisable to award gift cards based on this) but for many use cases where this is a bonus engagement thing will probably work fine.

The BasicPointsForBadges allows specifying a certain set of activities, each of which have a point total associated with them.  The current supported activities are:
- Connecting an external account in a set of categories
- Completing a specific survey
  - Can specify a limit to the number of times we would award points for this
- Hitting a certain level of daily data activity
  - For instance, 10 hours of wear time, 5,000 steps, etc.
- "Custom Activity" driven off of a custom field
  - e.g. if you want to have points driven off of a custom field based on any segment criteria you set up 

The data underlying this is all stored in a (writeable) JSON custom field.  Each activity has a corresponding "state" in the JSON, which stores the points it has awarded so far and any data it needs to ensure it does not award additional points for the same activity (e.g. the previously connected providers).  The state also stores the current badges as an array of point totals for when those badges were reached.

This design allows for some flexibility for changing point totals moving forward.  For instance:
- Lets say you wanted to award more points for connecting EHR; you can change the points awarded for connecting EHR and it will not recalculate all past points awarded for that connection
- Let's say you want to change badges to be awarded at every 2,000 points instead of 1,000; this will not "take away" any badges the user has previously earned, but it will instead just start counting points from the highest badge total earned (e.g. if badges were earned at [1000,2000], it'll just award the next badge at 4000 instead of 3000)

Of course, since this is in a custom field, we can clear out the custom field at any time if we want to "reset" a user's points and badges for them; or also edit the custom field JSON.

## Security

REMINDER: All file contents are public.

- [x] I have ensured no secure credentials or sensitive information remain in code, metadata, comments, etc.
  - [x] Please verify that you double checked that .storybook/preview.js does not contain your participant access key details. 
  - [x] There are no temporary testing changes committed such as API base URLs, access tokens, print/log statements, etc.
- [x] These changes do not introduce any security risks, or any such risks have been properly mitigated.

Describe briefly what security risks you considered, why they don't apply, or how they've been mitigated.

### Testing
- [ ] MyDataHelps iOS app
- [ ] MyDataHelps Android app
- [ ] [MyDataHelps website](mydatahelps.org)

### Documentation
- [x] I have added relevant Storybook updates to this PR as well.
- [ ] If this feature requires a developer doc update, tag @CareEvolution/api-docs.

Consider "Squash and merge" as needed to keep the commit history reasonable on `main`.

## Reviewers

Assign to the appropriate reviewer(s). Minimally, a second set of eyes is needed ensure no non-public information is published. Consider also including:
- Subject-matter experts
- Style/editing reviewers
- Others requested by the content owner

## Summary by CodeRabbit

- **New Features**
  - Introduced `BasicBadges` and `BasicPointsForBadges` components for displaying badges and points towards badges.
  - Added `DailyDataGoal` component to display daily data goals.
  - Added `NewBadgeView` component to show a detailed badge view with animations.
  - Implemented functionality for awarding points for connecting external accounts.

- **Style**
  - Added and updated styles for badges, points, daily data goals, and progress bars.
  - Introduced styling for the new `NewBadgeView` and `DailyDataGoal` components.

- **Localization**
  - Added new strings for points, badges, and Fitbit wear time in multiple languages, including German, English, Spanish, French, Italian, Dutch, and Portuguese.

### CODE REVIEW

**Comment by chrisnowak** [comment]
This should be ready for review now after a handful of iterations / tweaks

File: src/components/view/NewBadgeView/NewBadgeView.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                            {props.showBadgePoints &&
+                                <>{Math.max(...badges).toLocaleString()}<br />points</>
+                            }
+                        </>
+                    </BasicBadge>
+                </div>
+                <Title order={2} style={{ margin: "16px", marginBottom: 0 }}>{props.title || language("new-badge-title")}</Title>
+                <TextBlock>{props.text || language("new-badge-text")}</TextBlock>
+                <Button onClick={() => MyDataHelps.dismiss()}>
+                    {language("get-badge")}
```
Comment (reneefromhold) [inline src/components/view/NewBadgeView/NewBadgeView.tsx:59]: Bear with my questions, but does this show as a pop up? The MyDataHelps.dismiss is throwing me off.

File: src/components/view/NewBadgeView/NewBadgeView.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                <div className="mdhui-new-badge-view-badge-container">
+                    <BasicBadge size="xl" backgroundColor={getColorFromAssortment(badges.length - 1)}>
+                        <>
+                            {props.showBadgePoints &&
+                                <>{Math.max(...badges).toLocaleString()}<br />points</>
+                            }
+                        </>
+                    </BasicBadge>
+                </div>
+                <Title order={2} style={{ margin: "16px", marginBottom: 0 }}>{props.title || language("new-badge-title")}</Title>
```
Comment (reneefromhold) [inline src/components/view/NewBadgeView/NewBadgeView.tsx:56]: move styles to css

File: src/components/view/NewBadgeView/NewBadgeView.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import React, { useEffect } from "react";
```
Comment (reneefromhold) [inline src/components/view/NewBadgeView/NewBadgeView.tsx:1]: unused useEffect

File: src/components/container/BasicBadges/BasicBadges.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import React from "react";
+import { getColorFromAssortment, language, useInitializeView } from "../../../helpers";
+import { BasicBadge, Title } from "../../presentational";
+import "./BasicBadges.css"
+import MyDataHelps from "@careevolution/mydatahelps-js";
+import { parsePointsAndBadgesState } from "../../../helpers/BasicPointsAndBadges/Activities";
+import { FontAwesomeSvgIcon } from "react-fontawesome-svg-icon";
```
Comment (reneefromhold) [inline src/components/container/BasicBadges/BasicBadges.tsx:7]: remove unused

File: src/components/container/DailyDataGoal/DailyDataGoal.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    parameters: {
+        layout: 'fullscreen',
+    }
+};
+
+const render = (args: DailyDataGoalProps) => <Layout colorScheme='auto'>
+    <DailyDataGoal {...args} />
+</Layout>;
+
+export const SleepGoal = {
```
Comment (reneefromhold) [inline src/components/container/DailyDataGoal/DailyDataGoal.stories.tsx:20]: Are these stories meant to demonstrate a difference in presentation based on dailyDataType or how close someone was/not to their goal. I don't see a difference in the test data generated.
These look more like Goal Achieved, Goal Progress, and Goal No Progress.

File: src/components/presentational/BasicBadge/BasicBadge.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import type { ReactElement } from "react";
```
Comment (reneefromhold) [inline src/components/presentational/BasicBadge/BasicBadge.tsx:1]: Was no stories intentional?

File: src/components/presentational/Divider/Divider.tsx [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import React from "react";
+import "./Divider.css"
```
Comment (reneefromhold) [inline src/components/presentational/Divider/Divider.tsx:2]: No Stories?

**Review by reneefromhold — APPROVED** [review]
Code looks good. Just some general questions.

Tested against GarminActiveMinutes data and custom field content (needed to update below with the GarminActiveMinutes)
{"badges":[1000,2000,3000,4000,5000,6000],"activityStates":{"FitbitSleepData":{"pointsAwarded":3800,"daysAwarded":["2024-07-08","2024-07-09","2024-07-10","2024-07-11","2024-07-12","2024-07-14","2024-07-15","2024-07-16","2024-07-17","2024-07-18","2024-07-19","2024-07-20","2024-07-21","2024-07-22","2024-07-23","2024-07-24","2024-07-25","2024-07-26","2024-07-27"]},"FitbitWearTime":{"pointsAwarded":1900,"daysAwarded":["2024-07-09","2024-07-08","2024-07-10","2024-07-11","2024-07-12","2024-07-13","2024-07-14","2024-07-15","2024-07-16","2024-07-17","2024-07-18","2024-07-19","2024-07-20","2024-07-21","2024-07-22","2024-07-23","2024-07-24","2024-07-25","2024-07-26"]},"FitbitOrder":{"countCompleted":1,"lastQueryDate":"2024-07-09T23:16:37.77Z","pointsAwarded":800},"ConnectEhr":{"pointsAwarded":0}}}

File: src/components/view/NewBadgeView/NewBadgeView.tsx (hunk 1 above)
Comment (chrisnowak) [reply src/components/view/NewBadgeView/NewBadgeView.tsx:59]: Yep, this view is meant to be shown as a modal popup

File: src/components/container/DailyDataGoal/DailyDataGoal.stories.tsx (hunk 5 above)
Comment (chrisnowak) [reply src/components/container/DailyDataGoal/DailyDataGoal.stories.tsx:20]: Basically just exercising a real-world example.  Sleep in this case someone gets credit for even if they just had 1 minute of sleep recorded; whereas wear time only counts if you hit 10 hours etc.

File: src/components/container/BasicPointsForBadges/BasicPointsForBadges.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            customFields: {
+                [props.customField]: JSON.stringify({
+                    badges: [props.pointsPerBadge, props.pointsPerBadge * 2],
+                    activityStates: previewActivityStates
+                })
+            }
+        }
+    }
+
+    async function previewAwardPointsAndBadges(
```
[the remark this answers is not in the document]
Comment (chrisnowak) [reply src/components/container/BasicPointsForBadges/BasicPointsForBadges.tsx]: Done, good call

File: src/components/container/BasicPointsForBadges/BasicPointsForBadges.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            await previewAwardPointsAndBadges(props.activities, currentState);
+        if (!props.previewState) {
+            await persistPointsAndBadgesState(props.customField, updatedState);
+        }
+        let newPointTotal = sumActivityPoints(updatedState.activityStates);
+        setPoints(newPointTotal);
+
+        if (currentState.badges.length < updatedState.badges.length) {
+            await new Promise(resolve => setTimeout(resolve, 3000));
+            MyDataHelps.openApplication(props.awardBadgesViewUrl, { modal: true });
```
[the remark this answers is not in the document]
Comment (chrisnowak) [reply src/components/container/BasicPointsForBadges/BasicPointsForBadges.tsx:51]: Yep, there's a popup that shows them their new badge

File: src/components/view/NewBadgeView/NewBadgeView.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import { BasicBadge, Button, Layout, LoadingIndicator, TextBlock, Title } from "../../presentational";
+import "./NewBadgeView.css";
+import { parsePointsAndBadgesState } from "../../../helpers/BasicPointsAndBadges/Activities";
+import { startConfetti, stopConfetti } from '../../../helpers/confetti';
+
+export interface NewBadgeViewProps {
+    title?: string;
+    text?: string;
+    colorScheme?: 'auto' | 'light' | 'dark';
+    primaryColor?: string;
```
[the remark this answers is not in the document]
Comment (chrisnowak) [reply src/components/view/NewBadgeView/NewBadgeView.tsx:13]: Yes!  Good catch

File: src/components/presentational/BasicBadge/BasicBadge.tsx (hunk 6 above)
Comment (chrisnowak) [reply src/components/presentational/BasicBadge/BasicBadge.tsx:1]: Added stories

File: src/components/presentational/Divider/Divider.tsx (hunk 7 above)
Comment (chrisnowak) [reply src/components/presentational/Divider/Divider.tsx:2]: Added story

**Pull request merged**